modules/financial_bot/financial_bot/chains.py

```diff
@@ -0,0 +1,82 @@
+from typing import Any, Dict, List
+
+import qdrant_client
+from financial_bot.embeddings import EmbeddingModelSingleton
+from financial_bot.template import PromptTemplate
+from langchain.chains.base import Chain
+from langchain.llms import HuggingFacePipeline
+
+
+class ContextExtractorChain(Chain):
+    """
+    Encode the question, search the vector store for top-k articles and return
+    context news from documents collection of Alpaca news.
+    """
+
+    top_k: int = 1
+    embedding_model: EmbeddingModelSingleton
+    vector_store: qdrant_client.QdrantClient
+    vector_collection: str
+    output_key: str = "payload"
+
+    @property
+    def input_keys(self) -> List[str]:
+        return ["about_me", "question", "context"]
+
+    @property
+    def output_keys(self) -> List[str]:
+        return [self.output_key]
+
+    def _call(self, inputs: Dict[str, Any]) -> Dict[str, Any]:
+        # TODO: handle that None, without the need to enter chain
```

Comment:
Should this method be called `__call__` instead of `_call` ? I don't know the exact mechanics of LangChain, so it is just a question.

Reply:
It looks like that's the definition in langchain custom `Chain` module.
![Screenshot from 2023-10-03 13-42-43](https://github.com/iusztinpaul/hands-on-llms/assets/23241108/73c6f380-a3a3-43a6-9e73-de03a6c923e9)